File: 18-4sum/4sum.js
var fourSum = function(nums, target) {
    let lookup = new Set()
    let output = []
    let arr = [...nums].sort((a, b)=> a - b)

    for (let i = 0; i < nums.length - 3; i++) {
        for (let j = i + 1; j < nums.length - 2; j++) {
            first = arr[i]
            second = arr[j]

            a = j + 1
            b = nums.length - 1

            while (a < b) {
                sum = first + second + arr[a] + arr[b]

                if (sum == target) {
                    quadruplets = first + "," + second + "," + arr[a] + "," + arr[b]
                    if (!(lookup.has(quadruplets))) {
                        lookup.add(quadruplets)
                        output.push([first, second, arr[a], arr[b]])
                    }
                    a += 1
                    b -=1
                }
                else if (sum < target) a += 1
                else b -= 1
            }

        }
    }  
    return output
};

File: 18-4sum/4sum.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        lookup = set()
        output = []
        nums.sort()

        for a in range(len(nums) - 3):
            for b in range(a + 1, len(nums) - 2):
                first = nums[a]
                second = nums[b]

                c = b + 1
                d = len(nums) - 1

                while (c < d):
                    sum = first + second + nums[c] + nums[d]
                    quad = (first, second, nums[c], nums[d])

                    if (sum == target):
                        if (quad not in lookup):
                            output.append([first, second, nums[c], nums[d]])
                        lookup.add(quad)
                        c += 1
                        d -= 1
                    elif (sum < target):
                        c += 1
                    else:
                        d -= 1

        return output


        return result
        